src/hook.js

```diff
@@ -0,0 +1,78 @@
+
+/**
+ * @typedef {function} HookedFunction
+ * @property {function(function(), [number])} addHook A method that takes a new function to attach as a hook
+ *  to the HookedFunction
+ * @property {function(function())} removeHook A method to remove attached hooks
+ */
+
+/**
+ * A map of global hook methods to allow easy extension of hooked functions that are intended to be extended globally
+ * @type {{}}
+ */
+export const hooks = {};
+
+/**
+ * A utility function for allowing a regular function to be extensible with additional hook functions
+ * @param {string} type The method for applying all attached hooks when this hooked function is called
+ * @param {function()} fn The function to make hookable
+ * @param {string} hookName If provided this allows you to register a name for a global hook to have easy access to
+ *  the addHook and removeHook methods for that hook (which are usually accessed as methods on the function itself)
+ * @returns {HookedFunction} A new function that implements the HookedFunction interface
+ */
+export function createHook(type, fn, hookName) {
+  let _hooks = [{fn, priority: 0}];
+
+  let types = {
+    sync: function(...args) {
+      _hooks.forEach(hook => {
+        hook.fn.apply(this, args);
+      });
+    },
+    asyncSeries: function(...args) {
```

Comment:
I'm not sure exactly what the issue(s) are... but I uncovered some weird bugs in `asyncSeries` hooks.

The following test passes, as expected:

```javascript
    it('should allow context to be passed to hooks, but keep bound contexts', () => {
      let context;
      let fn = function() {
        context = this;
      };

      let boundContext1 = {};
      let calledBoundContext1;
      let hook1 = function(next) {
        calledBoundContext1 = this;
        next()
      }.bind(boundContext1);

      let hookFn = createHook('asyncSeries', fn);
      hookFn.addHook(hook1);

      hookFn();

      expect(calledBoundContext1).to.equal(boundContext1);
    });
```

However, it begins failing if you add a second hook, like so:

```javascript
   it('should allow context to be passed to hooks, but keep bound contexts', () => {
      let context;
      let fn = function() {
        context = this;
      };

      let boundContext1 = {};
      let calledBoundContext1;
      let hook1 = function(next) {
        calledBoundContext1 = this;
        next()
      }.bind(boundContext1);

      let boundContext2 = {};
      let calledBoundContext2;
      let hook2 = function(next) {
        calledBoundContext2 = this;
        next()
      }.bind(boundContext2);

      let hookFn = createHook('asyncSeries', fn);
      hookFn.addHook(hook1);

      // let newContext = {};
      // hookFn = hookFn.bind(newContext);
      hookFn();

      // expect(context).to.equal(newContext);
      expect(calledBoundContext1).to.equal(boundContext1);
      expect(calledBoundContext2).to.equal(boundContext2);
    });
```

It also fails with a single hook if the hookable function is bound:

```javascript
    it('should allow context to be passed to hooks, but keep bound contexts', () => {
      let context;
      let fn = function() {
        context = this;
      };

      let boundContext1 = {};
      let calledBoundContext1;
      let hook1 = function(next) {
        calledBoundContext1 = this;
        next()
      }.bind(boundContext1);

      let hookFn = createHook('asyncSeries', fn);
      hookFn.addHook(hook1);

      let newContext = {};
      hookFn = hookFn.bind(newContext);
      hookFn();

      expect(context).to.equal(newContext);
      expect(calledBoundContext1).to.equal(boundContext1);
    });
```

The error messages aren't very helpful... just `TypeError: Cannot read property 'replace' of undefined`, with no line number.

Reply:
The first test case you list I think is failing because you never added the second hook.  I think adding `hookFn.addHook(hook2);` should fix it.

The second test case is a bug.  I wasn't properly maintaining the `this` reference in the `asyncSeriesNext` function.  I updated the cod to use a fat-arrow function to keep the proper `this` and updated the test w/ your use case and it seems to be passing now.  Thanks.